src/main/java/mc/duzo/persona/common/battle/data/ServerBattleData.java

```diff
@@ -0,0 +1,207 @@
+package mc.duzo.persona.common.battle.data;
+
+import mc.duzo.persona.PersonaMod;
+import mc.duzo.persona.common.battle.BattleHandler;
+import mc.duzo.persona.common.battle.ai.BattleAI;
+import mc.duzo.persona.common.battle.turn.BattleTurn;
+import mc.duzo.persona.common.battle.turn.ServerBattleTurn;
+import mc.duzo.persona.common.skill.Skill;
+import mc.duzo.persona.network.PersonaMessages;
+import mc.duzo.persona.util.AbsoluteBlockPos;
+import mc.duzo.persona.util.DeltaTimeManager;
+import mc.duzo.persona.util.PersonaUtil;
+import net.minecraft.entity.Entity;
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.nbt.NbtCompound;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.network.ServerPlayerEntity;
+import net.minecraft.server.world.ServerWorld;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+public class ServerBattleData extends BattleData {
+	private AbsoluteBlockPos battlePos;
+	private ServerBattleTurn turn;
+
+	public ServerBattleData(List<ServerPlayerEntity> players, List<LivingEntity> targets, AbsoluteBlockPos pos) {
+		super();
+
+		for (ServerPlayerEntity player : players) {
+			this.players.add(player.getUuid());
+		}
+		for (LivingEntity target : targets) {
+			this.targets.add(target.getUuid());
+		}
+
+		this.battlePos = pos;
+	}
+	public ServerBattleData(List<ServerPlayerEntity> players, List<LivingEntity> targets) {
+		this(players,targets,null);
+
+		if (players.isEmpty()) return;
+		ServerPlayerEntity found = players.get(0);
+
+		this.battlePos = new AbsoluteBlockPos(found.getBlockPos(), found.getServerWorld());
+	}
+	public ServerBattleData(NbtCompound nbt) {
+		super(nbt);
+	}
+
+	public void toClient(ServerPlayerEntity target) {
+		PersonaMessages.syncBattleData(target, this);
+	}
+	public void toClient() {
+		for (PlayerEntity player : this.getPlayers()) {
+			this.toClient((ServerPlayerEntity) player);
+		}
+		for (LivingEntity target : this.getTargets()) {
+			if (target instanceof ServerPlayerEntity) {
+				this.toClient((ServerPlayerEntity) target);
+			}
+		}
+	}
+
+	@Override
+	protected boolean shouldUpdateCache() {
+		return DeltaTimeManager.isOnDelay(this.getCacheKey());
+	}
+
+	@Override
+	public BattleTurn getTurn() {
+		if (this.turn == null) {
+			this.turn = new ServerBattleTurn(this);
+		}
+
+		return this.turn;
+	}
+
+	public ServerBattleTurn getServerTurn() {
+		return (ServerBattleTurn) this.getTurn();
+	}
+
+	private void createCacheDelay() {
```

Comment:
Instead of repeating these cache methods, make a separate class for it (or subclass idrk)